### PR TITLE
[Extension] Add azmirror cloud endpoint for installing extensions in air gapped clouds and support config of index url

### PR DESCRIPTION
**Pull request by fengzhou-msft**
**Description**  

To support installing extensions in air-gapped clouds, we have set up a [pipeline](https://dev.azure.com/azure-sdk/internal/_build?definitionId=1608) to sync the extension index, wheel packages and extension command tree to a public storage account called `azmirror`. A Teleport process will be set up to sync everything in `azmirror` to different storage accounts inside air-gapped clouds. 

In order to use those files in each air-gapped cloud with Azure CLI, this PR adds an `azmirror_storage_account_resource_id` endpoint whose value may look like `https://azmirroragc.blob.core.windows.net`.
CLI will get extension related files from the endpoint if it is set in the current cloud, otherwise it will use the current default values (for public cloud and sovereign clouds).

This PR also adds config option `extension.index_url` to support customized extension index url. For example, to use the `index.json` on github:
`az config set extension.index_url=https://raw.githubusercontent.com/Azure/azure-cli-extensions/master/src/index.json`

**Testing Guide**  

```
$ az cloud update --cloud-config "{\"endpoint_azmirror_storage_account_resource_id\":\"https://azmirror.blob.core.windows.net\"}"
$ az cloud show  # you should see the value set for azmirrorStorageAccountResourceId
$ az extension list-available -otable --debug  # you should see a request sent to https://azmirror.blob.core.windows.net
$ az config set extension.index_url=https://raw.githubusercontent.com/Azure/azure-cli-extensions/master/src/index.json
$ az extension list-available -otable --debug  # Now you should see a request sent to raw.githubusercontent.com
```
**History Notes**  

[Extension] Support config of extension index url

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Extension

File: src/azure-cli-core/azure/cli/core/extension/_resolve.py [hunk 1]
```diff
@@ -88,6 +88,13 @@ def resolve_from_index(extension_name, cur_version=None, index_url=None, target_
     download_url, digest = chosen.get('downloadUrl'), chosen.get('sha256Digest')
     if not download_url:
         raise NoExtensionCandidatesError("No download url found.")
+    ext_endpoint = cli_ctx.cloud.endpoints.extension_storage_account_resource_id if cli_ctx and \
+        cli_ctx.cloud.endpoints.has_endpoint_set('extension_storage_account_resource_id') else None
+    config_index_url = cli_ctx.config.get('extension', 'index_url', None) if cli_ctx else None
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/extension/_resolve.py:93]: Maybe we can put these 2 URLs in one place?

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/extension/_resolve.py:93]: Also, isn't the `whl` location hard-coded in [the index](https://raw.githubusercontent.com/Azure/azure-cli-extensions/master/src/index.json)?

```
{
    "downloadUrl": "https://azurecliprod.blob.core.windows.net/cli-extensions/account-0.2.0-py2.py3-none-any.whl",
    "filename": "account-0.2.0-py2.py3-none-any.whl",
    "metadata": {
        "azext.isExperimental": true,
```

**Comment by yungezz** [comment]
put this to backlog now given environment setup is under process. When process done, will merge the PR.

File: src/azure-cli-core/azure/cli/core/parser.py [hunk 2]
```diff
@@ -298,11 +298,16 @@ def _get_extension_command_tree(self):
...
         EXT_CMD_TREE.load(os.path.join(cli_ctx.config.config_dir, 'extensionCommandTree.json'), VALID_SECOND)
         if not EXT_CMD_TREE.data:
+            import posixpath
             import requests
             from azure.cli.core.util import should_disable_connection_verify
             try:
+                ext_endpoint = cli_ctx.cloud.endpoints.extension_storage_account_resource_id if cli_ctx and \
+                    cli_ctx.cloud.endpoints.has_endpoint_set('extension_storage_account_resource_id') else None
+                url = posixpath.join(ext_endpoint, 'extensionCommandTree.json') if ext_endpoint else \
+                    'https://aka.ms/azExtCmdTree'
```
Comment (fengzhou-msft) [inline src/azure-cli-core/azure/cli/core/parser.py:308]:
```suggestion
                azmirror_endpoint = cli_ctx.cloud.endpoints.azmirror_storage_account_resource_id if cli_ctx and \
                    cli_ctx.cloud.endpoints.has_endpoint_set('azmirror_storage_account_resource_id') else None
                url = posixpath.join(azmirror_endpoint, 'extensions', 'extensionCommandTree.json') if azmirror_endpoint else \
                    'https://aka.ms/azExtCmdTree'
```

File: src/azure-cli-core/azure/cli/core/parser.py [hunk 3]
```diff
@@ -298,11 +298,16 @@ def _get_extension_command_tree(self):
...
         EXT_CMD_TREE.load(os.path.join(cli_ctx.config.config_dir, 'extensionCommandTree.json'), VALID_SECOND)
         if not EXT_CMD_TREE.data:
+            import posixpath
             import requests
             from azure.cli.core.util import should_disable_connection_verify
             try:
+                azmirror_endpoint = cli_ctx.cloud.endpoints.azmirror_storage_account_resource_id if cli_ctx and \
+                    cli_ctx.cloud.endpoints.has_endpoint_set('azmirror_storage_account_resource_id') else None
+                url = posixpath.join(azmirror_endpoint, 'extensions', 'extensionCommandTree.json') if azmirror_endpoint else \
+                    'https://aka.ms/azExtCmdTree'
```
Comment (fengzhou-msft) [inline src/azure-cli-core/azure/cli/core/parser.py:308]:
```suggestion
                url = posixpath.join(azmirror_endpoint, 'extensions', 'extensionCommandTree.json') if \
                    azmirror_endpoint else 'https://aka.ms/azExtCmdTree'
```

File: src/azure-cli-core/azure/cli/core/extension/_resolve.py (hunk 1 above)
Comment (fengzhou-msft) [reply src/azure-cli-core/azure/cli/core/extension/_resolve.py]: The index will be teleported into different storage accounts in air gapped clouds, we do not have permission to modify the whl `downloadUrl` based on the destination storage accounts after it gets copied in air gapped clouds. So we need to modify the `downloadUrl` here in azure-cli when we need to download it.

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/extension/_resolve.py]: Do we really need to put one URL in cloud config, one in `az config`?

Comment (fengzhou-msft) [reply src/azure-cli-core/azure/cli/core/extension/_resolve.py]: The one in cloud config is now a general support for downloading extensions and other tools/files needed in air gapped clouds. It's always a storage account endpoint and controlled by Microsoft.

The one in `az config` is to support customers that want to use their own customized `index.json`, which may have their own private extensions and some allowed public extensions. You can use any url of the index file, such as the [raw content url](https://raw.githubusercontent.com/Azure/azure-cli-extensions/master/src/index.json) on a github branch.

**Review by jiasli — APPROVED** [review]

**Review by msyyc — APPROVED** [review]

**Pull request merged**